### PR TITLE
Cleanup of SecretStore.

### PR DESCRIPTION
**Pull request by chriseth**
Added documentation, more failure checking and some general cleanup.

### CODE REVIEW

**Comment by LianaHus** [comment]
compile errors

**Comment by chriseth** [comment]
Compile errors should be fixed now.

File: libdevcore/CommonIO.cpp [hunk 1]
```diff
@@ -119,9 +119,11 @@ string dev::contentsString(std::string const& _file)
 	return ret;
 }
 
-void dev::writeFile(std::string const& _file, bytesConstRef _data)
+bool dev::writeFile(std::string const& _file, bytesConstRef _data)
```
Comment (arkpar) [inline libdevcore/CommonIO.cpp]: why not throw exception?

**Comment by chriseth** [comment]
I will change this to use exceptions.

**Comment by chriseth** [comment]
Updated.

File: libdevcrypto/SecretStore.cpp [hunk 2]
```diff
@@ -84,36 +85,38 @@ static js::mValue upgraded(std::string const& _s)
...
 }
 
-h128 SecretStore::importSecret(bytes const& _s, std::string const& _pass)
+h128 SecretStore::importSecret(bytes const& _s, string const& _pass)
 {
-	h128 r = h128::random();
-	m_cached[r] = _s;
-	m_keys[r] = make_pair(encrypt(_s, _pass), std::string());
-	save();
+	h128 r;
```
Comment (gavofyork) [inline libdevcrypto/SecretStore.cpp]: not better to throw in case difficulty to encrypt rather than return a zero uuid?

**Comment by gavofyork** [comment]
also, 4 platforms failing?

**Comment by chriseth** [comment]
I just wanted to see if I can get it to succeed the test here. We should probably have a second thought about how and when the nonce is written to the file and what to do in the failure case. Also note that it is written in the destructor, and we should catch the exception there.

**Comment by chriseth** [comment]
First wanted to see if the last commit succeeds the tests. We should probably have a second thought about the nonce file. Also note that the None destructor may throw...

File: libdevcrypto/Exceptions.h [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+#include <libdevcore/Exceptions.h>
+
+namespace dev
+{
+namespace crypto
+{
+
+/// Rare malfunction of cryptographic functions.
+DEV_SIMPLE_EXCEPTION_RLP(CryptoException);
```
Comment (gavofyork) [inline libdevcrypto/Exceptions.h]: should be `DEV_SIMPLE_EXCEPTION`, no?

File: libdevcrypto/Common.cpp [hunk 4]
```diff
@@ -259,12 +280,12 @@ h256 Nonce::get(bool _commit)
 			BOOST_THROW_EXCEPTION(InvalidState());
 		
 		// prevent seed reuse if process terminates abnormally
-		writeFile(s_seedFile, bytes());
+		try { writeFile(s_seedFile, bytes()); } catch (FileError const&) {}
```
Comment (subtly) [inline libdevcrypto/Common.cpp]: As per comment on previous line, this needs to throw. If it crashes: perfect! There's no sense or use in hiding something like this.

File: libdevcrypto/Common.cpp [hunk 5]
```diff
@@ -259,12 +280,12 @@ h256 Nonce::get(bool _commit)
...
 	}
 	h256 prev(s_seed);
 	sha3(prev.ref(), s_seed.ref());
 	if (_commit)
-		writeFile(s_seedFile, s_seed.asBytes());
+		try { writeFile(s_seedFile, s_seed.asBytes()); } catch (FileError const&) {}
```
Comment (subtly) [inline libdevcrypto/Common.cpp]: Same as above. If this is throwing the user *must* know about it. At the very least we get a crash report.

File: libdevcrypto/SecretStore.cpp [hunk 6]
```diff
@@ -183,62 +190,66 @@ h128 SecretStore::readKeyContent(std::string const& _content, std::string const&
...
 }
 
-std::string SecretStore::encrypt(bytes const& _v, std::string const& _pass, KDF _kdf)
+static bytes deriveNewKey(string const& _pass, KDF _kdf, js::mObject& o_ret)
 {
-	js::mObject ret;
-
-	// KDF info
 	unsigned dklen = 32;
+	unsigned iterations = 1 << 19;
```
Comment (subtly) [inline libdevcrypto/SecretStore.cpp]: 1 << 18

Comment (chriseth) [inline libdevcrypto/SecretStore.cpp]: Thanks!

File: libdevcore/CommonIO.cpp [hunk 7]
```diff
@@ -121,7 +121,10 @@ string dev::contentsString(std::string const& _file)
 
 void dev::writeFile(std::string const& _file, bytesConstRef _data)
 {
-	ofstream(_file, ios::trunc|ios::binary).write((char const*)_data.data(), _data.size());
+	ofstream s(_file, ios::trunc | ios::binary);
+	s.write(reinterpret_cast<char const*>(_data.data()), _data.size());
+	if (!s)
```
Comment (CJentzsch) [inline libdevcore/CommonIO.cpp]: shouldn't that be checked before trying to write to the file?
Additionally using a try/catch block?

File: libdevcrypto/SecretStore.cpp [hunk 8]
```diff
@@ -29,6 +29,7 @@
 #include <libdevcore/SHA3.h>
 #include <libdevcore/FileSystem.h>
 #include <test/JsonSpiritHeaders.h>
```
Comment (CJentzsch) [inline libdevcrypto/SecretStore.cpp]: > [Mittwoch, 17. Juni 2015 11:56:01 gavofyork] Yeah we should move over to libjsoncpp completely at some point. 
Not required for this PR, only nice to have.

File: libdevcrypto/SecretStore.cpp [hunk 9]
```diff
@@ -84,35 +86,37 @@ static js::mValue upgraded(std::string const& _s)
 	return js::mValue();
 }
 
-SecretStore::SecretStore(std::string const& _path): m_path(_path)
+SecretStore::SecretStore(string const& _path): m_path(_path)
 {
-	load();
+	if (!m_path.empty())
```
Comment (CJentzsch) [inline libdevcrypto/SecretStore.cpp]: Shouldn't we throw otherwise, since load is necessary for other operations?

File: libdevcrypto/SecretStore.cpp [hunk 10]
```diff
@@ -132,50 +136,53 @@ void SecretStore::clearCache() const
...
+			fs::remove(filename);
 	}
 }
 
-void SecretStore::load(std::string const& _keysPath)
+void SecretStore::load(string const& _keysPath)
 {
 	fs::path p(_keysPath);
-	boost::filesystem::create_directories(p);
+	fs::create_directories(p);
```
Comment (CJentzsch) [inline libdevcrypto/SecretStore.cpp]: needs a check whether it is successful or not.

File: libdevcrypto/SecretStore.h [hunk 11]
```diff
@@ -35,41 +35,81 @@ enum class KDF {
...
+ * UUID of the key.
+ * @note that most of the functions here affect the filesystem and throw exceptions on failure,
+ * and they also throw exceptions upon rare malfunction in the cryptographic functions.
+ */
 class SecretStore
 {
 public:
+	/// Construct a new SecretStore and read all keys in the given directory.
 	SecretStore(std::string const& _path = defaultPath());
-	~SecretStore();
```
Comment (CJentzsch) [inline libdevcrypto/SecretStore.h]: why did you remove the destructor?

Comment (chriseth) [inline libdevcrypto/SecretStore.h]: I don't think it is necessary, or did I miss something?

File: libdevcore/CommonIO.cpp (hunk 7 above)
Comment (chriseth) [inline libdevcore/CommonIO.cpp]: It is checked internally in `ofstream::write`. Furthermore, `ofstream` functions should not usually throw exceptions, they only set flags which can be checked by using e.g. `!s`.

File: libdevcrypto/SecretStore.cpp (hunk 9 above)
Comment (chriseth) [inline libdevcrypto/SecretStore.cpp]: It is not strictly necessary. If no path is provided (which is not the case in real use of this class), it just does not store the keys permanently.

File: libdevcrypto/SecretStore.cpp (hunk 10 above)
Comment (chriseth) [inline libdevcrypto/SecretStore.cpp]: It throws on failure.

File: libdevcrypto/SecretStore.cpp (hunk 9 above)
Comment (subtly) [inline libdevcrypto/SecretStore.cpp]: Please remove this condition and let create_directories throw. I don't see how the condition is necessary; if it is, consider a ensureFileStore() method which calls create_directories and is called by constructor and save(). This check is inevitable so I can't see a reason to defer the call. If load() works here, then save() *should* work later. This class is useless if it can't ensure read/write to m_path; thus, a good idea to check/ensure in the constructor.

File: libdevcore/CommonIO.cpp (hunk 7 above)
Comment (chfast) [inline libdevcore/CommonIO.cpp]: Not bad, but `if (s.fail())` would be more informative.

File: libdevcrypto/SecretStore.cpp [hunk 12]
```diff
@@ -132,50 +136,53 @@ void SecretStore::clearCache() const
...
 		v["crypto"] = crypto;
 		v["id"] = uuid;
 		v["version"] = c_keyFileVersion;
 		writeFile(filename, js::write_string(js::mValue(v), true));
-		if (!k.second.second.empty() && k.second.second != filename)
-			boost::filesystem::remove(k.second.second);
-		k.second.second = filename;
+		swap(k.second.filename, filename);
+		if (!filename.empty() && !fs::equivalent(filename, k.second.filename))
+			fs::remove(filename);
```
Comment (subtly) [inline libdevcrypto/SecretStore.cpp]: Suggestion for future: boost vs write && rename on success. Some file systems act differently when fs errors occur (disk space, permissions, etc.). And unfortunately, due to default partitioning schemes, dual-boot scenarios, etc., running out of disk space is something which happens a *LOT* (among thousands of users). Doing something similar to vim would save someone in both of these situations: write-out _file.json, delete file.json, then rename _file.json to file.json. If something bad happens in the middle it's recoverable.

File: libdevcrypto/SecretStore.h [hunk 13]
```diff
@@ -35,41 +35,81 @@ enum class KDF {
...
+	/// Store all keys in the managed directory.
 	void save() { save(m_path); }
 
+	/// @returns the default path for the managed directory.
 	static std::string defaultPath() { return getDataDir("web3") + "/keys"; }
 
 private:
+	struct EncryptedKey
+	{
+		std::string encryptedKey;
```
Comment (subtly) [inline libdevcrypto/SecretStore.h]: Curious. Why string and not bytes?

File: libdevcrypto/SecretStore.cpp [hunk 14]
```diff
@@ -84,35 +86,37 @@ static js::mValue upgraded(std::string const& _s)
...
 	return key;
 }
 
-h128 SecretStore::importSecret(bytes const& _s, std::string const& _pass)
+h128 SecretStore::importSecret(bytes const& _s, string const& _pass)
 {
-	h128 r = h128::random();
+	h128 r;
+	EncryptedKey key{encrypt(_s, _pass), string()};
+	if (!key.encryptedKey.empty())
```
Comment (subtly) [inline libdevcrypto/SecretStore.cpp]: !empty() ??

File: libdevcrypto/SecretStore.cpp [hunk 15]
```diff
@@ -84,35 +86,37 @@ static js::mValue upgraded(std::string const& _s)
...
+		BOOST_THROW_EXCEPTION(crypto::CryptoException());
```
Comment (subtly) [inline libdevcrypto/SecretStore.cpp]: Why isn't this thrown inside encrypt?

File: libdevcrypto/SecretStore.cpp [hunk 16]
```diff
@@ -183,62 +190,66 @@ h128 SecretStore::readKeyContent(std::string const& _content, std::string const&
...
-bool SecretStore::recode(h128 const& _uuid, string const& _newPass, std::function<std::string()> const& _pass, KDF _kdf)
+bool SecretStore::recode(h128 const& _uuid, string const& _newPass, function<string()> const& _pass, KDF _kdf)
 {
-//	cdebug << "recode:" << toUUID(_uuid);
 	bytes s = secret(_uuid, _pass, true);
 	if (s.empty())
 		return false;
-	m_keys[_uuid].first = encrypt(s, _newPass, _kdf);
+	m_cached.erase(_uuid);
+	m_keys[_uuid].encryptedKey = encrypt(s, _newPass, _kdf);
```
Comment (subtly) [inline libdevcrypto/SecretStore.cpp]: If exception is not moved into encrypt then an exception is necessary here. If encrypt fails the key is lost.

File: libdevcrypto/SecretStore.h (hunk 13 above)
Comment (chriseth) [inline libdevcrypto/SecretStore.h]: I don't think I changed this here.

File: libdevcrypto/SecretStore.cpp (hunk 14 above)
Comment (chriseth) [inline libdevcrypto/SecretStore.cpp]: I guess I will also add some further tests for this class...

File: libdevcrypto/SecretStore.cpp (hunk 12 above)
Comment (chriseth) [inline libdevcrypto/SecretStore.cpp]: Added that feature to `writeFile`.

**Comment by chriseth** [comment]
Updated and added tests.

File: libdevcrypto/SecretStore.cpp (hunk 16 above)
Comment (CJentzsch) [inline libdevcrypto/SecretStore.cpp]: You added an error handling in the `srypt` function in the form of an throwing exception. I think this is a good place to catch and handle it.

File: test/libdevcrypto/crypto.cpp [hunk 17]
```diff
@@ -45,6 +49,7 @@ static CryptoPP::OID s_curveOID(CryptoPP::ASN1::secp256k1());
 static CryptoPP::DL_GroupParameters_EC<CryptoPP::ECP> s_params(s_curveOID);
 static CryptoPP::DL_GroupParameters_EC<CryptoPP::ECP>::EllipticCurve s_curve(s_params.GetCurve());
 
+
```
Comment (CJentzsch) [inline test/libdevcrypto/crypto.cpp]: uneeded free line

File: test/libdevcrypto/SecretStore.cpp [hunk 18]
```diff
@@ -29,11 +29,16 @@
...
+#include <test/TestHelper.h>
+#include <test/TestUtils.h>
 using namespace std;
 using namespace dev;
+using namespace dev::test;
 
 namespace js = json_spirit;
+namespace fs = boost::filesystem;
+
+BOOST_GLOBAL_FIXTURE( MoveNonceToTempDir );
```
Comment (CJentzsch) [inline test/libdevcrypto/SecretStore.cpp]: extra `;` -> build failure

File: libdevcrypto/SecretStore.cpp (hunk 16 above)
Comment (chriseth) [inline libdevcrypto/SecretStore.cpp]: The exceptions that are thrown there are actually kind of failed assertions, they should not really happen. What is more likely to happen are io errors and they are also not caught here.
So basically, this function returns false if the password was wrong and throws an exception in exceptional circumstances.
